heat/core/manipulations.py

```diff
@@ -699,6 +700,113 @@ def hstack(tup):
     return concatenate(tup, axis=axis)
 
 
+def reshape(a, shape):
+    """
+    Returns a tensor with the same data and number of elements as a, but with the specified shape.
+
+    NOTE: Only tensors with split = None and distributed vectors (dim = 1) are supported yet
+
+    Parameters
+    ----------
+    a : ht.DNDarray
+        The input tensor
+    shape : tuple
+        Shape of the new tensor
+
+    Returns
+    -------
+    reshaped : ht.DNDarray
+        The tensor with the specified shape
+
+    Examples
+    --------
+    >>> a = ht.zeros((3,4))
+    >>> ht.reshape(a, (4,3))
+    tensor([[0,0,0],
+            [0,0,0],
+            [0,0,0],
+            [0,0,0]])
+
+    >>> a = ht.linspace(0, 14, 8, split=0)
+    >>> ht.reshape(a, (2,4))
+    (1/2) tensor([[0., 2., 4., 6.]])
+    (2/2) tensor([[ 8., 10., 12., 14.]])
+    """
+
+    def reshape_counts_displs(csum1, csum2):
+        """
+        Calculate the counts and displacements needed for redistributing the data from two cumulative sums.
+        """
+        p = csum1[a.comm.rank - 1] if a.comm.rank > 0 else 0
+        i = 0
+
+        result = np.zeros((a.comm.size,), dtype=np.int)
+
+        # Find starting position with binary search
+        minpos, maxpos, = 0, a.comm.size
+
+        while True:
+            i = (maxpos - minpos) // 2 + minpos
+
+            if minpos == maxpos:
+                break
+
+            if csum2[i] < p:
+                minpos = i + 1
+            else:
+                maxpos = i
+
+        # write counts
+        while csum2[i] < csum1[a.comm.rank]:
+            result[i] = csum2[i] - p
+            p = csum2[i]
+            i += 1
+        result[i] = csum1[a.comm.rank] - p
+
+        return tuple(result), (0,) + tuple(np.cumsum(result[:-1]))
+
+    # Check the type of shape and number elements
+    if not isinstance(shape, (list, tuple)):
+        raise RuntimeError("shape must be a tuple or list")
+    if np.prod(shape) != a.size:
+        raise ValueError("cannot reshape array of size {} into shape {}".format(a.size, shape))
+
+    # Forward to Pytorch directly
+    if a.split is None:
+        return factories.array(
+            torch.reshape(a._DNDarray__array, shape), dtype=a.dtype, device=a.device, comm=a.comm
+        )
+
+    # Check for currently not supported tensors
+    if a.split > 0:
+        raise NotImplementedError("Tensors with split axes > 0 are not supported yet")
+
+    # Create new flat result tensor
+    _, local_shape, _ = a.comm.chunk(shape, a.split)
+    data = torch.empty(np.prod(local_shape), dtype=a.dtype.torch_type())
```

Comment:
device call, and numpy doesnt run on GPU

Reply:
Thanks for the hint. I hadn't the opportunity to test it on the GPU.
I don't really get the second remark. np.prod will return an integer?

Reply:
It runs on GPU. DNDarray handles wrong devices, but, nevertheless, it is better to change it beforehand.

Reply:
the numpy command will run on the CPU process even if the device is GPU, which is why we need to call the torch version. it isnt going to save a lot of time but out programming model is to put it all on one device.

so it would be:

```python
_, local_shape, _ = a.comm.chunk(shape, a.split)
lshp = torch.prod(torch.tensor(local_shape, device=a.device.torch_device), dim=0)
data = torch.empty(local_shape, dtype=a.dtype.torch_type(), device=a.device.torch_device)
```
im not sure if torch requires the dim argument for prod

Reply:
If you say so, I will change it :confused: 
Does it really save time? It seems more like an overhead for me creating a tensor and pushing data to the gpu and back for the calculation.

Reply:
with the use of torch it does not leave the gpu. when you use numpy and the device is gpu then it has to leave the gpu, go to the cpu, then go back to the gpu after it's done.